independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,72 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
+
+    private volatile boolean releasing = false;
```

Comment:
No need for this: you need to use compareAndSet loops on both retain/release to make sure that you cannot "resurrect" a released resource

eg 
- ref count start at 1 i.e. alive
- a release put it to 0 (and could have rights to close the file)
- an acquire just find it at 0, returning `false` because the resource is already released

Reply:
I'm using this `boolean` for a twofold reason:

1. checking that a release hasn't been already requested when `readLockAcquire` is called because in that case I cannot return a reference to the `jarFile` that could have been closed in the meanwhile.
2. checking if a release has been already requested when `readLockRelease` is called: in this case if the readers counter reaches to 0 I'm actually allowed to close the `jarFile`.

I don't think that I can reuse the `readersCounter` for this same purpose, for instance setting it to 0 when a release is requested, because doing so I would lose track of how many readers are still currently accessing the `jarFile`. Am I missing something?

Reply:
I think this is still introducing a level of freedom which can break things: we should always relies on a single source of truth related the state of the file ownership; and the math which bring 1 ->  0 as a final state requires such, while the boolean one could be used as an hint (but at that point you can implement the same with an `isReleased() { return rf.cnt() ==  0; } ` having the same API.

Reply:
> The boolean hint could be re-implemented with an `isReleased() { return rf.cnt() == 0; } ` still granting a nice external API.

The 2 things are unrelated though: the fact that the counter is zero (there are no current readers) doesn't imply in any way that somebody call a close on the `JarResource` thus actually asking a release of the `jarFile`.

I understand that ideally it would be much better to have the state of the `JarFileReference`, but still I don't see how to achieve it in this case. Moreover I tried to (mentally) double check all the possible interactions and situations and I believe that it shouldn't be any gap. Please review it if you have time.

Reply:
Looking at the happens before relation twice I cannot see evident problems (so this reduce my concerns), but i personally prefer to reduce the level of freedoms to make more easy to reason of things, but is personal, really.